### PR TITLE
Fix unhelpful type error

### LINKED ISSUE #54: Unhelpful error with extra arguments to an init
## Checklist

- [x] I updated to the latest version and the problem persists.
- [x] I searched the [existing GitHub issues][1] and this hasn't already been reported.

## Overview

When an init inside of a `Service` closure has an extra argument, often the error shows up at the very top of the `register` block, rather than at the problematic service.

## Example Reproduction

```swift
protocol P {}
class C: P {
    init(a: Int, c: String) {}
}

func setup() {
    Factory.register {
        Service(.transient, P.self) { r in
            C(a: r.resolve(), b: r.resolve(), c: r.resolve())
        }
    }
}
```

## Behavior

The error should show up for the service in question -- ideally pointing out the extraneous argument `b`.

## Environment

| Software    | Version |
|:------------|:--------|
| macOS       | `12.6.0` |
| Xcode       | `14.0.1` |
| Dependiject | `0.3.0` |

[1]: https://github.com/Tiny-Home-Consulting/Dependiject/issues?q=is%3Aissue

### PR DESCRIPTION
**Pull request by bbrk24**
## Issue Link

Fixes #54 

Jira issue: [THOS-5](https://tinyhomeconsultingllc.atlassian.net/browse/THOS-5)

## Overview of Changes

This PR fixes type resolution errors. Basically, what would happen in a situation like this:

```swift
protocol P {}
class C: P {
    init(a: Int, c: String) {}
}

func setup() {
    Factory.register { // [1]
        Service(.transient, P.self) { r in
            C(
                a: r.resolve(),
                b: r.resolve(), // [2]
                c: r.resolve()
            )
        }
    }
}
```

is that Swift would be unable to figure out the generic type argument to `r.resolve()` at [2], since the initializer for `C` doesn't take a `b`. Because it can't determine the return type, swift would report "type of expression is ambiguous without more context", completely ignoring the fact that it's not meant to be there. Due to the result builder transform, the error would be reported at [1] instead of [2].

Adding an overload that returns `Any` works around this problem: if it can't figure out how to call `resolve<T>() -> T`, it tries the other overload, `resolve() -> Any`. This allows it to complete type-checking and realize that `b` shouldn't even be there.

### Anything you want to highlight?

I marked the overload as `@available(*, unavailable)`. This has two consequences:

1. In a line like `var x: Any = Factory.shared.resolve()`, this is inferred as `resolve<Any>()` instead of the new overload.
2. A line with no types at all, such as `var x = Factory.shared.resolve()`, still emits an error: 
Screen Shot 2022-10-13 at 9 56 12 AM

However, it has the downside that the new overload doesn't show up in the documentation. Using `deprecated` instead of `unavailable` makes it show up in the documentation, but also makes point 1 no longer true, and downgrades point 2 from an error to a warning.

## Test Plan

Paste the code snippet above (or any equivalent) into a project that imports Dependiject. You should see the error at [2] instead of [1].

### CODE REVIEW

File: Dependiject/Resolver.swift [hunk 1]
```diff
@@ -33,4 +33,18 @@ public extension Resolver {
...
         return self.resolve(T.self, name: name)
     }
+    
+    /// This overload fixes compiler errors.
+    ///
+    /// To explicitly specify the type, use ``resolve(_:name:)`` or ``resolve(_:)`` instead.
+    ///
+    /// Without this overload, adding extraneous arguments to an `init` inside of the callback
+    /// passed to ``Service/init(_:_:name:_:)`` does not error at the `init` call, but at the top
+    /// of the ``Factory/register(builder:)`` block.
```
Comment (Ljgleeson) [inline Dependiject/Resolver.swift:43]: Might be worth adding a note mentioning that this overload doesn't appear in the documentation just incase someone is confused why it isn't.

Comment (bbrk24) [reply Dependiject/Resolver.swift:43]: Sure, that makes sense to me.

**Review by NicholasChase — APPROVED** [review]

**Review by Ljgleeson — APPROVED** [review]

File: Dependiject/Resolver.swift [hunk 2]
```diff
@@ -19,7 +19,7 @@ public extension Resolver {
     /// Get an unnamed dependency of the specified type.
     /// - Returns: The unnamed instance of the type.
     /// - Important: There must be an unnamed registration of the specified type. When resolving a
-    /// named dependency, use ``resolve(_:name:)`` or ``resolve(name:)`` instead.
+    /// named dependency, use ``resolve(_:name:)`` or ``Resolver/resolve(name:)`` instead.
```
Comment (wboyd600) [inline Dependiject/Resolver.swift:22]: did this link not work?

Comment (bbrk24) [reply Dependiject/Resolver.swift:22]: It used to, but for some reason when I added the overload it didn't anymore, and I had to specify the base type.

File: Dependiject/Resolver.swift [hunk 3]
```diff
@@ -33,4 +33,21 @@ public extension Resolver {
...
+    ///
+    /// Without this overload, adding extraneous arguments to an `init` inside of the callback
+    /// passed to ``Service/init(_:_:name:_:)`` does not error at the `init` call, but at the top
+    /// of the ``Factory/register(builder:)`` block.
+    ///
+    /// - Remark: This overload does not show up in the documentation because it is explicitly
+    /// marked `unavailable`.
+    @available(*, unavailable, renamed: "resolve(_:name:)", message: """
+        If this is being called, Swift cannot determine the correct type.
+        """)
```
[the remark this answers is not in the document]
Comment (bbrk24) [reply Dependiject/Resolver.swift:49]: Yeah I don't know what's going on here.

[the remark this answers is not in the document]
Comment (bbrk24) [reply Dependiject/Resolver.swift:49]: I just tested this in Xcode 13.2.1 to see if it has the same problem, but it ended up being worse: it changed from `resolve` to `resolveresolve`.

[the remark this answers is not in the document]
Comment (wboyd600) [reply Dependiject/Resolver.swift:49]: maybe we should just make this 
```
    @available(*, unavailable, message: """
        Dependencies can only be resolved implicitly inside of the Factory.register() function.
        """)
```

I don't think the XCode bug is going away anytime soon and I'd hate to replace one unhelpful error with another

[the remark this answers is not in the document]
Comment (bbrk24) [reply Dependiject/Resolver.swift:49]: That message isn't true, though. Just because we haven't been doing it in views doesn't mean we can't:

```swift
struct ExampleView: View {
    var viewModel: ExampleViewModel = Factory.shared.resolve()

    var body: some View { ... }
}
```

[the remark this answers is not in the document]
Comment (wboyd600) [reply Dependiject/Resolver.swift:49]: Let me try again: 

maybe we should just make this
```swift
    @available(*, unavailable, message: """
        {Something true and helpful}
        """)
```
I don't think the XCode bug is going away anytime soon and I'd hate to replace one unhelpful error with another
